### PR TITLE
[libc++] <experimental/simd> Add compound assignment operators for simd reference

### PR DESCRIPTION
**Pull request by joy2myself**

### CODE REVIEW

**Comment by joy2myself** [comment]
@philnik777 Hi! This and the other two patches (https://github.com/llvm/llvm-project/pull/86478, https://github.com/llvm/llvm-project/pull/88091) are ready for review.

**Comment by joy2myself** [comment]
@philnik777 Gentle ping.

**Comment by joy2myself** [comment]
Hi @philnik777,

Just following up on this PR and #88091. Could you please take a look when you get a chance?

I have some follow-up patches ready, but they may depend on these currently open PRs. It would be great if these two PRs could be merged soon.

Thanks!

File: libcxx/include/experimental/__simd/reference.h [hunk 1]
```diff
@@ -60,6 +63,24 @@ class __simd_reference {
...
+  _LIBCXX_SIMD_REFERENCE_OP_(-)
+  _LIBCXX_SIMD_REFERENCE_OP_(*)
+  _LIBCXX_SIMD_REFERENCE_OP_(/)
+  _LIBCXX_SIMD_REFERENCE_OP_(%)
+  _LIBCXX_SIMD_REFERENCE_OP_(&)
+  _LIBCXX_SIMD_REFERENCE_OP_(|)
+  _LIBCXX_SIMD_REFERENCE_OP_(^)
+  _LIBCXX_SIMD_REFERENCE_OP_(<<)
+  _LIBCXX_SIMD_REFERENCE_OP_(>>)
+#  undef _LIBCXX_SIMD_REFERENCE_OP_
```
Comment (philnik777) [inline libcxx/include/experimental/__simd/reference.h:82]: I think we should just expand this. Macros result in notoriously bad error messages and here it really doesn't seem that bad to just expand the macros.

File: libcxx/test/std/experimental/simd/simd.reference/reference_operators.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  template <class T, class SimdAbi>                                                                                    \
+  struct SimdReferenceOperatorHelper##name {                                                                           \
+    template <class U>                                                                                                 \
+    void operator()() const {                                                                                          \
+      ex::simd<T, SimdAbi> origin_simd(static_cast<T>(3));                                                             \
+      static_assert(noexcept(origin_simd[0] op## = static_cast<U>(2)));                                                \
+      origin_simd[0] op## = static_cast<U>(2);                                                                         \
+      assert((T)origin_simd[0] == (T)(static_cast<T>(3) op static_cast<T>(std::forward<U>(2))));                       \
+    }                                                                                                                  \
+  };
```
Comment (philnik777) [inline libcxx/test/std/experimental/simd/simd.reference/reference_operators.pass.cpp:40]: Could we refactor this to take an operation as a functor instead? So the instantiation looks something like `SimdReferenceOperatorHelper<T, SimdAbi, decltype([](auto& lsh, auto ths) { lhs += rhs; })>` etc.

File: libcxx/include/experimental/__simd/reference.h [hunk 3]
```diff
@@ -60,6 +63,24 @@ class __simd_reference {
     return {__s_, __idx_};
   }
 
+#  define _LIBCXX_SIMD_REFERENCE_OP_(__op)                                                                             \
+    template <class _Up, class = decltype(std::declval<value_type&>() __op## = std::declval<_Up>())>                   \
```
Comment (philnik777) [inline libcxx/include/experimental/__simd/reference.h:67]: Isn't this just `is_assignable`?

**Review by philnik777 — CHANGES_REQUESTED** [review]
Sorry for the slow review, I didn't have a lot of time recently.

**Comment by joy2myself** [comment]
@philnik777 Hello, I have removed all macro definitions according to your comment. However, the compiler does not allow lambda expressions to be passed directly into the template, so I use struct to wrap lambda expressions to pass template parameters in test files.

In addition, due to the error of insufficient memory or flash capacity in ARMv7 CI, I split the test cases into two test files.

Any other comments?

File: libcxx/include/experimental/__simd/reference.h [hunk 4]
```diff
@@ -71,6 +72,86 @@ class __simd_reference {
 
   template <class _Tp1, class _Storage1, class _Vp1>
   friend void swap(__simd_reference<_Tp1, _Storage1, _Vp1>&& __a, _Vp1& __b) noexcept;
+
+  template <
+      class _Up,
+      enable_if_t<is_assignable_v<value_type&, decltype(std::declval<value_type&>() + std::declval<_Up>())&&>, int> = 0>
```
Comment (philnik777) [inline libcxx/include/experimental/__simd/reference.h:78]: I think this should be reverted to the previous state. I misread what this did before, sorry.

**Review by philnik777 — APPROVED** [review]

**Pull request merged**